### PR TITLE
8271203: C2: assert(iff->Opcode() == Op_If || iff->Opcode() == Op_CountedLoopEnd || iff->Opcode() == Op_RangeCheck) failed: Check this code when new subtype is added

### PR DESCRIPTION
**Pull request by kelthuzadx**
Hi, I'm trying to fix [JDK-8271203](https://bugs.openjdk.java.net/browse/JDK-8271203). It's reasonable to unswitch LongCountedLoop, so relax it.

image

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271203](https://bugs.openjdk.java.net/browse/JDK-8271203): C2: assert(iff->Opcode() == Op_If || iff->Opcode() == Op_CountedLoopEnd || iff->Opcode() == Op_RangeCheck) failed: Check this code when new subtype is added

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to c2d47b2e14717b4a5bca2f0561876bc6eb7c43d4
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to c2d47b2e14717b4a5bca2f0561876bc6eb7c43d4
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4920/head:pull/4920` \
`$ git checkout pull/4920`

Update a local copy of the PR: \
`$ git checkout pull/4920` \
`$ git pull https://git.openjdk.java.net/jdk pull/4920/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4920`

View PR using the GUI difftool: \
`$ git pr show -t 4920`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4920.diff">https://git.openjdk.java.net/jdk/pull/4920.diff</a>

</details>

### CODE REVIEW

**Review by chhagedorn — APPROVED** [review]
I don't think that that a `LongCountedLoop` is unswitched here looking at the stack trace but it's rather chosen as unswitch `If`. I had a closer look at it. It seems that the loop
```
for (long l1 = i; l1 < 2; ++l1) {
     iArrFld[0] += 5;
}
```
is peeled once and then the original loop is found to be dead (because it is known that the type of the iv `i` of the outer-most loop is >= 1). From this peeled iteration we only have `403 LongCountedLoopEnd` left which now just acts as a normal `If` node. It looks fine to then simply use this as an unswitch if for unswitching `367 CountedLoop`. So, relaxing the assert to allow a `LongCountedLoopEnd` node in `dominated_by` looks good IMO.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

File: test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8271203
+ * @summary C2: assert(iff->Opcode() == Op_If || iff->Opcode() == Op_CountedLoopEnd || iff->Opcode() == Op_RangeCheck) failed: Check this code when new subtype is added
+ * @library /test/lib
+ * @run main/othervm -Xbatch -XX:-TieredCompilation -XX:CompileOnly=UnswitchingLongCountedLoop compiler.c2.UnswitchingLongCountedLoop
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java:30]: Isn't that CompileOnly statement missing a the package name or a wildcard?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java (hunk 1 above)
Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java:30]: And `@library /test/lib` is not needed.

**Comment by kelthuzadx** [comment]
> I don't think that a `LongCountedLoop` is unswitched here looking at the stack trace but it's rather chosen as unswitch `If`. I had a closer look at it. It seems that the loop
> 
> ```
> for (long l1 = i; l1 < 2; ++l1) {
>      iArrFld[0] += 5;
> }
> ```
> 
> is peeled once and then the original loop is found to be dead (because it is known that the type of the iv `i` of the outer-most loop is >= 1). From this peeled iteration we only have `403 LongCountedLoopEnd` left which now just acts as a normal `If` node. It looks fine to then simply use this as an unswitch if for unswitching `367 CountedLoop`. So, relaxing the assert to allow a `LongCountedLoopEnd` node in `dominated_by` looks good IMO.

Hi Christian, I think you are right. As far as I see, after some loop optimizations, loop at line-17 and loop at line-19 are no longer real loops after PhaseIdealLoop2:

image

I'm not sure why there are two same loops(B9 and B19-B22) .. but anyway, 403#LongCountedLoopEnd is located in B21, it's not a loop anymore since no backedge found and C2 treats it as a unswitch_iff.

File: test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java (hunk 1 above)
Comment (kelthuzadx) [reply test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java:30]: Hi Tobias, I've removed `@library /test/lib`, also according to  @chhagedorn 's analysis, I think `LongCountedLoopAsUnswitchIff` should be a better test name.

**Comment by kelthuzadx** [comment]
Thanks @chhagedorn @rwestrel and @TobiHartmann for reviews.

File: test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java (hunk 1 above)
Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java]: But the `CompileOnly` statement is still missing the package name, right? No compilations will happen.

Comment (kelthuzadx) [reply test/hotspot/jtreg/compiler/c2/UnswitchingLongCountedLoop.java]: Oops...Added

**Review by TobiHartmann — APPROVED** [review]
Looks good to me but please verify that the test reproduces the issue without your fix.

**Comment by kelthuzadx** [comment]
> Looks good to me but please verify that the test reproduces the issue without your fix.

Yes, I will test if it would crash without this fix before integrating.

**Comment by chhagedorn** [comment]
> > I don't think that a `LongCountedLoop` is unswitched here looking at the stack trace but it's rather chosen as unswitch `If`. I had a closer look at it. It seems that the loop
> > ```
> > for (long l1 = i; l1 < 2; ++l1) {
> >      iArrFld[0] += 5;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is peeled once and then the original loop is found to be dead (because it is known that the type of the iv `i` of the outer-most loop is >= 1). From this peeled iteration we only have `403 LongCountedLoopEnd` left which now just acts as a normal `If` node. It looks fine to then simply use this as an unswitch if for unswitching `367 CountedLoop`. So, relaxing the assert to allow a `LongCountedLoopEnd` node in `dominated_by` looks good IMO.
> 
> Hi Christian, I think you are right. As far as I see, after some loop optimizations, loop at line-17 and loop at line-19 are no longer real loops after PhaseIdealLoop2:
> 
> I'm not sure why there are two same loops(B9 and B19-B22) .. but anyway, 403#LongCountedLoopEnd is located in B21, it's not a loop anymore since no backedge found and C2 treats it as a unswitch_iff.

Not sure which lines you refer to in your image. Lines in the JTreg test (they do not match)? And at what point is this image taken? Can you also elaborate more on what you mean by two same loops?

**Comment by kelthuzadx** [comment]
JTREG TEST:
```

public class Reduced {
    static int iArrFld[] = new int[400];

    public static void main(String[] strArr) {
        for (int i = 0; i < 10; i++) {
            test();
        }
    }

    static void test() {
        int i = 56, i2 = 22257;

        do {
            do {
                int i24 = 1;
                while (++i24 < 2) {
                }
                for (long l1 = i; l1 < 2; ++l1) {
                    iArrFld[0] += 5;
                }
            } while ((i2 -= 2) > 0);  //line 22
            for (long l3 = 8; l3 < 194; ++l3) { // line 23
            }
        } while (--i > 0);  //line25
    }
}
```
Well, the above CFG is corresponding to the last step(before beautify loops, [loop.xml.log](https://github.com/openjdk/jdk/files/6982404/loop.xml.log) ) before crash. I haven't learned about loopopts in C2 yet,  it looks like B9 and B19-22 are refer to the same loop(`do{...} while ((i2 -= 2) > 0); `), the above CFG looks like:
```
 static void test() {
        int i = 56, i2 = 22257;

        do {
             if (xx) {
               do{ ... } while ((i2 -= 2) > 0); 
             } else {
               do{ ... } while ((i2 -= 2) > 0); 
             }
            for (long l3 = 8; l3 < 194; ++l3) { // line 23
            }
        } while (--i > 0);  //line25
    }
```
...maybe that's not true.

**Comment by chhagedorn** [comment]
> Well, the above CFG is corresponding to the last step(before beautify loops, loop.xml.log ) before crash. I haven't learned about loopopts in C2 yet, it looks like B9 and B19-22 are refer to the same loop(do{...} while ((i2 -= 2) > 0); ), the above CFG looks like:

They probably appear to be very similar on first sight but are not. You can check the `-XX:+TraceLoopOpts` output to see what's going on with your loops. Maybe the loop was already unswitched once and now you have 2 very similar looking loops that correspond to the loop `do{...} while ((i2 -= 2) > 0); )` in the Java code. In general, there is not always a 1:1 relation between a loop in Java code and `LoopNode(s)` in the IR. For example, you could also be dealing with pre/main/post loops for a single Java loop which look very similar.

**Comment by kelthuzadx** [comment]
> > Well, the above CFG is corresponding to the last step(before beautify loops, loop.xml.log ) before crash. I haven't learned about loopopts in C2 yet, it looks like B9 and B19-22 are refer to the same loop(do{...} while ((i2 -= 2) > 0); ), the above CFG looks like:
> 
> They probably appear to be very similar on first sight but are not. You can check the `-XX:+TraceLoopOpts` output to see what's going on with your loops. Maybe the loop was already unswitched once and now you have 2 very similar looking loops that correspond to the loop `do{...} while ((i2 -= 2) > 0); )` in the Java code. In general, there is not always a 1:1 relation between a loop in Java code and `LoopNode(s)` in the IR. For example, you could also be dealing with pre/main/post loops for a single Java loop which look very similar.

Make sense. TraceLoopOpts may tell me everything about what happens. I’m curious about loop optimizations, this is a good start point to explore them:P

**Pull request closed without merging**